test/test-viz.cpp

```diff
@@ -0,0 +1,350 @@
+    bool check_inputs(Vertex& element, std::vector<std::string>& ins) {
+        for(Vertex& suggested_element : _scheme) {
+            for(int i = 0; i < ins.size(); i++) {
+                if(ins[i] == suggested_element.name) {
+                    element.ins.emplace_back(suggested_element);
+                    element.layer = suggested_element.layer + 1;
+                    ins.erase(ins.begin() + i);
+                    i--;
+                }
+            }
+        }
+
+        if(ins.empty()) return false;
+        return true;
+    }
+
+    //Method to handle elements with lacking bonds
+    void request_handling() {
+        while(!_queue.empty()) {
+            for (int i = 0; i < _queue.size(); i++) {
+                //Going through every element of the scheme to find ones that are in the queue and connecting them to the inputs
+                for (int j = 0; j < _scheme.size(); j++) {
+                    if (std::get<0>(_queue[i]).name == _scheme[j].name) {
+                        if (!check_inputs(_scheme[j], std::get<1>(_queue[i]))) {
+                            _queue.erase(_queue.begin() + i);
+                            i--;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    void output() {
+        for(auto vert : _scheme) {
+            std::cout << vert.type << ' ' << vert.name << ' ' << vert.layer << " INS: ";
+            for(auto ins : vert.ins) {
+                std::cout << ins.name << ' ';
+            }
+            std::cout << std::endl;
+        }
+        std::cout << "\n============\n";
+        for(auto requests : _queue) {
+            std::cout << std::get<0>(requests).type;
+        }
+    }
+
+    std::vector<Vertex> _scheme;
+    std::vector< std::tuple< Vertex, std::vector< std::string > > > _queue;
+    int _outputNum;
+};
+
+class bench_parser : public lorina::bench_reader {
+public:
+    explicit bench_parser(logic_scheme& logicScheme) : _scheme(logicScheme) {}
+
+    virtual void on_input(const std::string& name) const override {
+        _scheme.add_input(name);
+    }
+
+    virtual void on_output(const std::string& name) const override {
+        _scheme.add_output(name);
+    }
+
+    virtual void on_dff(const std::string &input, const std::string &output) const override {
+        _scheme.add_dff(input, output);
+    }
+
+    virtual void on_gate(const std::vector<std::string>& inputs, const std::string& output, const std::string& type) const override {
+        _scheme.add_gate(inputs, output, type);
+    }
+
+    logic_scheme& _scheme;
+};
+
+void buildRectangles(std::vector<Vertex>& elems, std::vector<SDL_FRect>& rects, std::vector<Line>& connections, int maxElem, double height, double width) {
+    if(elems.size() == maxElem) {
+        int count = 0;
+        for(auto elem : elems) {
+            SDL_FRect rect;
+            rect.x = elem.layer * 2 * width;
+            rect.y = count * 2 * height;
```

Comment:
"Magic" numbers (2 in this case) are not allowed in the source code. Use named constants in expressions or named functions to calculate the result.